File: 2024/clojure/src/day03.clj
(ns day03)

(defn parse-and-sum-muls [input]
  (let [pattern #"mul\((\d{1,3}),(\d{1,3})\)"
        matches (re-seq pattern input)]
    (reduce
     (fn [sum [_ x y]]
       (+ sum (* (Integer/parseInt x) (Integer/parseInt y))))
     0
     matches)))

(defn parse-and-sum-muls-with-conditions [input]
  (let [mul-pattern #"mul\((\d{1,3}),(\d{1,3})\)"
        do-pattern #"do\(\)"
        dont-pattern #"don't\(\)"
        tokens (re-seq (re-pattern (str mul-pattern "|" do-pattern "|" dont-pattern)) input)]
    (:sum
     (reduce
      (fn [state [token]]
        (cond
          (re-matches (re-pattern do-pattern) token)
          (assoc state :enabled true)

          (re-matches (re-pattern dont-pattern) token)
          (assoc state :enabled false)

          (re-matches (re-pattern mul-pattern) token)
          (let [[_ x y] (re-matches (re-pattern mul-pattern) token)]
            (if (:enabled state)
              (update state :sum + (* (Integer/parseInt x) (Integer/parseInt y)))
              state))))
      {:enabled true :sum 0}
      tokens))))

(let [answer (parse-and-sum-muls "xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))")]
  (println "Part 1 (Example):" answer))
(let [answer (parse-and-sum-muls-with-conditions "xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))")]
  (println "Part 2 (Example):" answer))

(let [answer (parse-and-sum-muls (slurp "data/day03.txt"))]
  (println "Part 1:" answer))
(let [answer (parse-and-sum-muls-with-conditions (slurp "data/day03.txt"))]
  (println "Part 2:" answer))
 

File: 2024/python/day03.py
import argparse
import re

from utils import read_lines

parser = argparse.ArgumentParser()
parser.add_argument("input", type=str, help="Path to input file")

mul_pattern = r"mul\((\d{1,3}),(\d{1,3})\)"
mul_regex = re.compile(mul_pattern)
do_pattern = r"do\(\)"
dont_pattern = r"don't\(\)"


def parse_and_sum_muls(x):
    matches = mul_regex.findall(x)
    return sum(int(match[0]) * int(match[1]) for match in matches)


def parse_and_sum_muls_with_conditions(x):
    pattern = re.compile(rf"{mul_pattern}|{do_pattern}|{dont_pattern}")
    tokens = [m[0] for m in re.finditer(pattern, x)]

    enabled = True
    total = 0

    for token in tokens:
        if re.fullmatch(do_pattern, token):
            enabled = True
        elif re.fullmatch(dont_pattern, token):
            enabled = False
        else:
            m = mul_regex.fullmatch(token)
            if m and enabled:
                x, y = int(m[1]), int(m[2])
                total += x * y

    return total


def part_1(x):
    return parse_and_sum_muls(x)


def part_2(x):
    return parse_and_sum_muls_with_conditions(x)


def main():
    args = parser.parse_args()
    x = "".join(read_lines(args.input))

    print("Part 1:", part_1(x))
    print("Part 2:", part_2(x))


if __name__ == "__main__":
    main()
